### PR TITLE
Non-record: EMA+SWA Tight Averaging with Fused TTT LoRA + Sliding Window (1.1371 BPB)

### PR DESCRIPTION
**Pull request by yunoshev**
## Non-record submission                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                       
  **val_bpb = 1.1371** (at 600s eval mark) | Artifact: 15.88 MB                                                                                                                                                                                        
                                                                                                                                                                                                                                                       
  11L D512 with EMA (0.997) + SWA tight averaging (no qgrid), fused TTT LoRA + sliding window eval, Muon 0.99, full GPTQ Hessians. Training: 5922 steps in 600s on 8xH100. Score at 600s eval mark: 1.1371 BPB.                                        
                                                                                                                                                                                                                                                       
  See `records/track_non_record_16mb/2026-04-05_EMA_SWA_TTT_LoRA_Fused_Sliding/README.md` for full details.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: EMA+SWA Tight Averaging with Fused TTT LoRA + Sliding Window (1.1371 BPB)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

## Analysis **PR #1366** — `records/track_non_record_16mb/2026-04-05_EMA_SWA_TTT_LoRA_Fused_Sliding/train_gpt.py` Head SHA: `d7f61f0aad38b34fdfa0be795ab0e780b70bc0f1` --- ### 1. N-gram / BigramHash: CLEAN The BigramHash (`bigram_hash_dim`) implementation at lines 571–573 XORs only input tokens: ``` bigram_idx[:,1:] = torch.bitwise_xor(36313*t[:,1:], 27191*t[:,:-1]) % bmod ``` The XOR is over `t[:,1:]` (current input token) and `t[:,:-1]` (previous input token). **Target tokens are never XOR'd into the hash key.** No ILLEGAL n-gram family bug present. The NgramCache (lines 12–56) uses dict-based context tuples derived from `x_flat` (context) and looks up `y_flat` (target) as a count key — standard legal use. The target is only used to look up a count, not to build the hash key. --- ### 2. TTT Pattern: LEGAL score-first per chunk (is_last_chunk guard) The function `eval_val_fused_ttt_sliding` (lines 195–307) implements a per-chunk loop: 1. **Score first**: For each chunk `ci`, the model runs `torch.inference_mode()` over all windows in that chunk (lines 245–253), accumulating NLL. Scoring happens **before** any training update for that chunk. 2. **Train after scoring, with is_last_chunk guard** (lines 254–281): ```python is_last_chunk = ci == num_chunks - 1 if not is_last_chunk and args.ttt_epochs > 0: # multi-epoch training on val_tokens[chunk_tok_start:chunk_tok_end] ``` The final chunk is **never trained on**. Earlier chunks are scored first, then used for TTT to improve the model for subsequent chunks. This matches the PR #1413 pattern. 3. **No Pre-Quant TTT**: The TTT call (line 1057–1060) happens **after** quantization roundtrip evaluation (`q_val_bpb` computed at line 1055). The model is adapted post-quantization, not as a multi-epoch pre-quant leakage pass. 4. The training data for TTT...

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
